consensus/types.go

```diff
@@ -450,73 +498,57 @@ func (t Transaction) SiafundOutputID(i int) SiafundOutputID {
-// the id of the SiafundOutput.
+// SiaClaimOutputID returns the ID of the SiacoinOutput that is created when
+// the siafund output is spent. The ID is the hash the SiafundOutputID.
 func (id SiafundOutputID) SiaClaimOutputID() SiacoinOutputID {
-	return SiacoinOutputID(crypto.HashAll(
-		id,
-	))
+	return SiacoinOutputID(crypto.HashObject(id))
 }
 
-// SigHash returns the hash of a transaction for a specific signature. `i` is
-// the index of the signature for which the hash is being returned. If
-// `WholeTransaction` is set to true for the signature, then all of the
-// transaction fields except the signatures are included in the transactions.
-// If `WholeTransaction` is set to false, then the fees, inputs, ect. are all
-// added individually. The signatures are added individually regardless of the
-// value of `WholeTransaction`.
+// SigHash returns the hash of the fields in a transaction covered by a given
+// signature. See CoveredFields for more details.
+//
+// TODO: make this take a CoveredFields instead?
 func (t Transaction) SigHash(i int) crypto.Hash {
-	cf := t.Signatures[i].CoveredFields
-	var signedData []byte
-	if cf.WholeTransaction {
-		signedData = encoding.MarshalAll(
-			t.SiacoinInputs,
-			t.SiacoinOutputs,
-			t.FileContracts,
-			t.FileContractTerminations,
-			t.StorageProofs,
-			t.SiafundInputs,
-			t.SiafundOutputs,
-			t.MinerFees,
-			t.ArbitraryData,
-			t.Signatures[i].InputID,
-			t.Signatures[i].PublicKeyIndex,
-			t.Signatures[i].Timelock,
-		)
-	} else {
-		for _, input := range cf.SiacoinInputs {
-			signedData = append(signedData, encoding.Marshal(t.SiacoinInputs[input])...)
-		}
-		for _, output := range cf.SiacoinOutputs {
-			signedData = append(signedData, encoding.Marshal(t.SiacoinOutputs[output])...)
-		}
-		for _, contract := range cf.FileContracts {
-			signedData = append(signedData, encoding.Marshal(t.FileContracts[contract])...)
-		}
-		for _, termination := range cf.FileContractTerminations {
-			signedData = append(signedData, encoding.Marshal(t.FileContractTerminations[termination])...)
-		}
-		for _, storageProof := range cf.StorageProofs {
-			signedData = append(signedData, encoding.Marshal(t.StorageProofs[storageProof])...)
-		}
-		for _, siafundInput := range cf.SiafundInputs {
-			signedData = append(signedData, encoding.Marshal(t.SiafundInputs[siafundInput])...)
-		}
-		for _, siafundOutput := range cf.SiafundOutputs {
-			signedData = append(signedData, encoding.Marshal(t.SiafundOutputs[siafundOutput])...)
-		}
-		for _, minerFee := range cf.MinerFees {
-			signedData = append(signedData, encoding.Marshal(t.MinerFees[minerFee])...)
-		}
-		for _, arbData := range cf.ArbitraryData {
-			signedData = append(signedData, encoding.Marshal(t.ArbitraryData[arbData])...)
-		}
+	// if WholeTransaction flag is set, hash everything but the signatures.
+	if t.Signatures[i].CoveredFields.WholeTransaction {
+		t.Signatures = []TransactionSignature{}
```

Comment:
We want to be able to sign signatures, it shouldn't be required to be empty if WholeTransaction is set.